### DISCUSSION TITLE
MAUI Droid Ntiers communication issue : Unable to read beyond the end of the stream

### DISCUSSION THREAD

**Discussion by sylum** [MarimerLLC/csla#3325] [Questions]
Hi Folks,

I am currently working on an MAUI project using CSLA6 and CSLA7Preview  and I am facing an issue with the ntiers communication between android and the appserver.
Actually, the maui app windows works well with the ntiers communication but with android i got the following dataportal error :
**Unable to read beyond the end of the stream**.

Below, the github link to reproduce the issue  base on the Rocky MauiExample projet
https://github.com/sylum/MauiExample_NTiers

![FZfqwzHtYa](https://user-images.githubusercontent.com/15035848/235188387-3202f57b-337f-4084-9d3c-6a7afec32e16.png)

**Exception :**
```
Csla.DataPortalException
  Message=Unable to read beyond the end of the stream.
  Source=Csla
  Arborescence des appels de procédure :

   at Csla.DataPortalClient.DataPortalProxy.Fetch(Type objectType, Object criteria, DataPortalContext context, Boolean isSync)
   at Csla.DataPortal`1.<DoFetchAsync>d__19[[BusinessLibrary.PersonList, BusinessLibrary, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].MoveNext()
   at Csla.DataPortal`1.<FetchAsync>d__24[[BusinessLibrary.PersonList, BusinessLibrary, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].MoveNext()
   at MauiExample.Pages.PersonListPage.<>c__DisplayClass3_0.<<RefreshData>b__0>d.MoveNext() in C:\RFit\Developpement\Csla\SOURCE\MauiExample\MauiExample\MauiExample\Pages\PersonListPage.xaml.cs:line 31
--- End of stack trace from previous location ---
   at Csla.Xaml.ViewModelBase`1.<RefreshAsync>d__69`1[[BusinessLibrary.PersonList, BusinessLibrary, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null],[BusinessLibrary.PersonList, BusinessLibrary, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].MoveNext()
   at MauiExample.Pages.PersonListPage.RefreshData() in C:\RFit\Developpement\Csla\SOURCE\MauiExample\MauiExample\MauiExample\Pages\PersonListPage.xaml.cs:line 35
   at MauiExample.Pages.PersonListPage.OnAppearing() in C:\RFit\Developpement\Csla\SOURCE\MauiExample\MauiExample\MauiExample\Pages\PersonListPage.xaml.cs:line 20
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_0(Object state)
   at Android.App.SyncContext.<>c__DisplayClass2_0.<Post>b__0() in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.App/SyncContext.cs:line 36
   at Java.Lang.Thread.RunnableImplementor.Run() in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Thread.cs:line 36
   at Java.Lang.IRunnableInvoker.n_Run(IntPtr jnienv, IntPtr native__this) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/obj/Release/net7.0/android-33/mcw/Java.Lang.IRunnable.cs:line 84
   at Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PP_V(_JniMarshal_PP_V callback, IntPtr jnienv, IntPtr klazz) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:line 22
```

Does someone has already encounter this issue ?

Thanks a lot for you feedback in advance,

Best regards,
Cedric

**Version and Platform**
CSLA : 6.2.2 and 7.0.0-R23042601
MAUI : WIN & DROID
VS2022
WINDOW11

**Comment by rockfordlhotka** [comment]
In my experience this often means that the server is returning a 500 or other unexpected result, so the data portal on the client isn't actually getting a result from the data portal on the server, but rather from the web server (or some other middleware).

If you are using the Android emulator, remember that it can't access your PCs localhost. I _think_ there might be ways around that? I usually just host the appserver in Azure so it has a public IP address to solve the issue.

**Comment by sylum** [comment]
Hi Rocky,

Thx for your answer. Actually, I'm using an Android Mobile (a real one) to access the appserver who is hosted by IIS on my PC.

In short, with the application running on Windows, I can have access to the appserver and everything works well but with the same application running on Android mobile,  i'm still keeping the error : **Unable to read beyond the end of the stream**.

However, the appserver can be reached by the android mobile (using the dataportal URL into chrome like below).

![Vysor_scgvgEYIwu](https://user-images.githubusercontent.com/15035848/235717120-2255d719-2308-4209-9158-4b7a47543d2e.png)

I feel like I'm missing something obvious. 
Does anyone have a idea or sample showing  MAUI N-tiers  on android.

Thanks a lot for you help,

Best regards,
Cedric

**Comment by sylum** [comment]
Just in case to complete my previous post, i put below the client dataportal configuration (droid)

![devenv_ccQMd38V0O](https://user-images.githubusercontent.com/15035848/235855858-88a9ebf7-1c64-4fc1-bdba-2f924d16fe95.png)

and below the appserver dataportal configuration

![devenv_khxvTJPauu](https://user-images.githubusercontent.com/15035848/235856043-77036966-51bb-4268-9600-e92bdd4b2a59.png)

**Reply by rockfordlhotka** [reply]
Looking at [HttpProxy](https://github.com/MarimerLLC/csla/blob/main/Source/Csla/DataPortalClient/HttpProxy.cs#L162), it catches the `WebException` and puts the stream contents into a `Message` field to help debugging.

It _might_ be that this exception with more detail is nested inside the top-level exception you are catching at the UI level. Can you see if there's an inner exception with more detail?

**Comment by sylum** [comment]
Hi Rocky,

I think a find something. Following your advice, I have tried to see what happens into the **HttpProxy** class.
Actually, the method **CallViaHttpClient** is called without exception, but next when the **DataPortalProxy** try to _Deserialize_ the data, the response came with an ErrorData **(Unable to read beyond the end of the stream**)

![image](https://user-images.githubusercontent.com/15035848/236208355-0afd0976-64d3-443a-8f9e-daa4100d9f16.png)

According the plateform (Win or Droid), I have noticed the serialized data seems different 

_Below Windows (working)_
![1](https://user-images.githubusercontent.com/15035848/236209227-bca4a147-9e50-4d6c-8ab9-3969a7932f77.png)

_Below Droid (not working)_
![2](https://user-images.githubusercontent.com/15035848/236209256-98878d8c-5482-4517-8620-9e18cc30e076.png)

I dont know if it's meaningful but if I'm forcing the dataportal to be isSync = True then the Droid works.

Thx a lot for your help.

**Reply by CROprogrammer** [reply]
Hi, I'm facing the same issue and I'm curious to know where shoud 'isSync = True' be put?

**Reply by CROprogrammer** [reply]
I've tried changing isSync to be True always but I'm still getting "Unable to read beyond the end of the stream." exception. I'm using CSLA 6.2.2 and Xamarin.
This exception does not occur when using AppServer & MVC or AppServer & WPF, only AppServer & Xamarin.

Exception occurs here:
```
   at System.IO.BinaryReader.InternalRead(Int32 numBytes)
   at Csla.Serialization.Mobile.CslaBinaryReader.Read(Stream serializationStream)
   at Csla.Serialization.Mobile.MobileFormatter.Deserialize(Stream serializationStream)
   at Csla.Server.Hosts.HttpPortalController.InvokePortal(String operation, Stream requestStream, Stream responseStream)
```

**Reply by sylum** [reply]
When debugging csla-6.2.2, I have spoted this point into DataPortalProxy.cs csla class.
By forcing the IsSync to True for the CallDataPortalServer method like below, the (maui) Android N-tiers works.

![image](https://github.com/MarimerLLC/csla/assets/15035848/4a61de93-11c7-4999-b451-969b7fcb7402)

**Reply by RolandoTonin** [reply]
I have the same problem. My Maui Android client App, calling an App Service on Azure throws the exception "Csla.DataPortalException: Unable to read beyond the end of the stream" every time I use FetchAsync or ExecuteAsync(). If instead I use Fetch() and Execute() everything works fine.
Both client and server side have CSLA 7.0.0 installed but the same problem I encountered with CSLA 6.2.2.
In the App Server's Program class I use:

builder.Services.AddCsla(o => o
     .AddAspNetCore()
     .DataPortal(dpo => dpo
     .EnableSecurityPrincipalFlowFromClient()
     .AddServerSideDataPortal()
     ));

while client side (MauiProgram class):
mauiAppBuilder.Services.AddCsla(o => o
                 .RegisterContextManager<Csla.Xaml.ApplicationContextManager>()
                 .DataPortal(dp => dp
                     .EnableSecurityPrincipalFlowFromClient()
                     .UseHttpProxy(hp =>
                     {
                         hp.DataPortalUrl = "https://XXX.azurewebsites.net/api/dataportal";
                     }
                     )));
if client side i built it for windows machine it works correctly. I experimented with the same code with a WPF App client and it works also fine.
The problem is with Android Emulator and also with a Samsung Android phone.
Thanks a lot for any suggestion
Rolando

**Comment by RolandoTonin** [comment]
hy all,
Have any of you had the same problem and found a solution?

thanks a lot
Rolando

**Comment by rockfordlhotka** [comment]
I have finally had time to replicate this issue so I can work on a solution. Issue tracking this is #3791

**Comment by rockfordlhotka** [comment]
I have a workaround at least.

If you configure the client and data portal endpoint (http portal controller) for `UseTextSerialization = true` then the data portal operates correctly.

I don't know yet if this is a CSLA bug or a limitation of Android (or MAUI?). The whole reason `UseTextSerialization` was added in the first place is because early versions of Blazor wasm couldn't do binary data transfers.

**Reply by kobi-shanti** [reply]
I can confirm I've experienced the same bug using Xamarin, so I don't believe it is restricted to MAUI.

**Reply by rockfordlhotka** [reply]
Maybe it has to do with Android?

At least we have a reasonable workaround.

**Reply by sylum** [reply]
Thx a lot, the workaround works perfectly.

**Comment by rockfordlhotka** [comment]
We now have a better workaround that allows you to continue to use the binary data transfer (which is more efficient).

Leave the data portal configuration as normal, and change the configuration of the `HttpClient` service in `MauiProgram.cs`:

```c#
  builder.Services.AddScoped<HttpClient>((p) => new HttpClient(new SocketsHttpHandler()));
```

https://github.com/MarimerLLC/csla/blob/main/docs/Upgrading%20to%20CSLA%208.md#maui-android
